src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,252 @@
+       exist more values for it, and hence we add it to terms. */
+    Node gstatus = valuation.getSatValue(g);
+    if (!gstatus.isNull() && gstatus.getConst<bool>())
+    {
+      terms.push_back(e);
+    }
+  }
+}
+
+bool CegisUnif::constructCandidates(const std::vector<Node>& enums,
+                                    const std::vector<Node>& enum_values,
+                                    const std::vector<Node>& candidates,
+                                    std::vector<Node>& candidate_values,
+                                    std::vector<Node>& lems)
+{
+  Assert(enums.size() == enum_values.size());
+  if (enums.empty())
+  {
+    return false;
+  }
+  unsigned min_term_size = 0;
+  std::vector<unsigned> enum_consider;
+  Trace("cegis-unif-enum") << "Register new enumerated values :\n";
+  for (unsigned i = 0, size = enums.size(); i < size; ++i)
+  {
+    Trace("cegis-unif-enum") << "  " << enums[i] << " -> " << enum_values[i]
+                             << std::endl;
+    unsigned sz = d_tds->getSygusTermSize(enum_values[i]);
+    if (i == 0 || sz < min_term_size)
+    {
+      enum_consider.clear();
+      min_term_size = sz;
+      enum_consider.push_back(i);
+    }
+    else if (sz == min_term_size)
+    {
+      enum_consider.push_back(i);
+    }
+  }
+  /* only consider the enumerators that are at minimum size (for fairness) */
+  Trace("cegis-unif-enum") << "...register " << enum_consider.size() << " / "
+                           << enums.size() << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
+  for (unsigned i = 0, ecsize = enum_consider.size(); i < ecsize; ++i)
+  {
+    unsigned j = enum_consider[i];
+    Node e = enums[j];
+    Node v = enum_values[j];
+    std::vector<Node> enum_lems;
+    d_sygus_unif.notifyEnumeration(e, v, enum_lems);
+    /* the lemmas must be guarded by the active guard of the enumerator */
+    Assert(d_enum_to_active_guard.find(e) != d_enum_to_active_guard.end());
+    Node g = d_enum_to_active_guard[e];
+    for (unsigned j = 0, size = enum_lems.size(); j < size; ++j)
+    {
+      enum_lems[j] = nm->mkNode(OR, g.negate(), enum_lems[j]);
+    }
+    lems.insert(lems.end(), enum_lems.begin(), enum_lems.end());
+  }
+  /* build candidate solution */
+  Assert(candidates.size() == 1);
+  Node vc = d_sygus_unif.constructSolution();
+  if (vc.isNull())
+  {
+    return false;
+  }
+  candidate_values.push_back(vc);
+  return true;
+}
+
+Node CegisUnif::purifyLemma(Node n,
+                            bool ensureConst,
+                            std::vector<Node>& model_guards,
+                            BoolNodePairMap& cache)
+{
+  unsigned i, size = n.getNumChildren();
+  Kind k = n.getKind();
+  bool fapp = k == APPLY_UF && size > 0 && n[0] == d_candidate;
+  Trace("cegis-unif-purify") << "PurifyLemma : " << n << "\n";
+  /* top level applications of function-to-synthsize should never be removed */
+  if (ensureConst || !fapp)
```

Comment:
I'm not sure why we are not caching if !ensureConst  && fapp.  I think this should be cached as well.